File: mini-services/qx-engine/src/engine/structure.ts
import type { Candle, StructureInfo, TrendDir } from '../types';

// ============ Market Structure Analysis ============
// EMA(9/21) trend + swing (HH/HL vs LH/LL) structure + pullback detection.
// This is the "মার্কেট স্ট্রাকচার" pillar of the confluence model.

function ema(values: number[], period: number): number {
  if (!values.length) return NaN;
  const k = 2 / (period + 1);
  let e = values[0];
  for (let i = 1; i < values.length; i++) e = values[i] * k + e * (1 - k);
  return e;
}

export function analyzeStructure(candles: readonly Candle[], pip: number): StructureInfo {
  const n = candles.length;
  const closes = candles.map((c) => c.close);
  const emaFast = ema(closes.slice(-40), 9);
  const emaSlow = ema(closes.slice(-60), 21);

  // swings over last ~80 candles
  const look = Math.min(n - 2, 80);
  const swingHighs: number[] = [];
  const swingLows: number[] = [];
  for (let i = Math.max(2, n - look); i < n - 2; i++) {
    const c = candles[i];
    let isHigh = true, isLow = true;
    for (let k = 1; k <= 2; k++) {
      if (candles[i - k].high >= c.high || candles[i + k].high >= c.high) isHigh = false;
      if (candles[i - k].low <= c.low || candles[i + k].low <= c.low) isLow = false;
    }
    if (isHigh) swingHighs.push(c.high);
    if (isLow) swingLows.push(c.low);
  }
  let swingTrend: TrendDir = 'RANGE';
  if (swingHighs.length >= 2 && swingLows.length >= 2) {
    const hh = swingHighs[swingHighs.length - 1] > swingHighs[swingHighs.length - 2];
    const hl = swingLows[swingLows.length - 1] > swingLows[swingLows.length - 2];
    const lh = swingHighs[swingHighs.length - 1] < swingHighs[swingHighs.length - 2];
    const ll = swingLows[swingLows.length - 1] < swingLows[swingLows.length - 2];
    if (hh && hl) swingTrend = 'UP';
    else if (lh && ll) swingTrend = 'DOWN';
  }

  const last = candles[n - 1];
  const emaGapPips = (emaFast - emaSlow) / pip;
  let trend: TrendDir = 'RANGE';
  if (emaGapPips > 1.2) trend = 'UP';
  else if (emaGapPips < -1.2) trend = 'DOWN';

  // strength: agreement of ema trend, swing trend, price vs ema
  let strength = 30;
  if (trend !== 'RANGE') strength += 25;
  if (swingTrend === trend && trend !== 'RANGE') strength += 25;
  const aboveFast = last.close > emaFast;
  if ((trend === 'UP' && aboveFast) || (trend === 'DOWN' && !aboveFast)) strength += 20;

  // pullback: within last 4 candles price dipped into ema zone then recovered (for UP)
  const last4 = candles.slice(-4);
  let pullback = false;
  if (trend === 'UP') {
    const dipped = last4.some((c) => c.low <= emaSlow + 2 * pip || c.low <= emaFast);
    pullback = dipped && last.close > emaFast && last.close > last.open;
  } else if (trend === 'DOWN') {
    const spiked = last4.some((c) => c.high >= emaSlow - 2 * pip || c.high >= emaFast);
    pullback = spiked && last.close < emaFast && last.close < last.open;
  }

  // consecutive candles
  let consecutiveBull = 0, consecutiveBear = 0;
  for (let i = n - 1; i >= Math.max(0, n - 10); i--) {
    if (candles[i].close > candles[i].open) consecutiveBull++;
    else break;
  }
  for (let i = n - 1; i >= Math.max(0, n - 10); i--) {
    if (candles[i].close < candles[i].open) consecutiveBear++;
    else break;
  }

  return {
    trend,
    emaFast,
    emaSlow,
    swingTrend,
    strength: Math.min(100, strength),
    atrPips: 0, // filled by caller
    consecutiveBull,
    consecutiveBear,
    pullback,
  };
}

File: mini-services/qx-engine/qxengine/structure.py
"""Market structure analysis — exact port of the TS structure.ts.

EMA(9/21) trend + swing (HH/HL vs LH/LL) structure + pullback detection.
"""


def _ema(values, period: int):
    if not values:
        return float("nan")
    k = 2.0 / (period + 1)
    e = values[0]
    for i in range(1, len(values)):
        e = values[i] * k + e * (1 - k)
    return e


def analyze_structure(candles, pip: float) -> dict:
    n = len(candles)
    closes = [c["close"] for c in candles]
    ema_fast = _ema(closes[-40:], 9)
    ema_slow = _ema(closes[-60:], 21)

    # swings over last ~80 candles
    look = min(n - 2, 80)
    swing_highs = []
    swing_lows = []
    for i in range(max(2, n - look), n - 2):
        c = candles[i]
        is_high = True
        is_low = True
        for k in (1, 2):
            if candles[i - k]["high"] >= c["high"] or candles[i + k]["high"] >= c["high"]:
                is_high = False
            if candles[i - k]["low"] <= c["low"] or candles[i + k]["low"] <= c["low"]:
                is_low = False
        if is_high:
            swing_highs.append(c["high"])
        if is_low:
            swing_lows.append(c["low"])

    swing_trend = "RANGE"
    if len(swing_highs) >= 2 and len(swing_lows) >= 2:
        hh = swing_highs[-1] > swing_highs[-2]
        hl = swing_lows[-1] > swing_lows[-2]
        lh = swing_highs[-1] < swing_highs[-2]
        ll = swing_lows[-1] < swing_lows[-2]
        if hh and hl:
            swing_trend = "UP"
        elif lh and ll:
            swing_trend = "DOWN"

    last = candles[n - 1]
    ema_gap_pips = (ema_fast - ema_slow) / pip
    trend = "RANGE"
    if ema_gap_pips > 1.2:
        trend = "UP"
    elif ema_gap_pips < -1.2:
        trend = "DOWN"

    strength = 30
    if trend != "RANGE":
        strength += 25
    if swing_trend == trend and trend != "RANGE":
        strength += 25
    above_fast = last["close"] > ema_fast
    if (trend == "UP" and above_fast) or (trend == "DOWN" and not above_fast):
        strength += 20

    last4 = candles[-4:]
    pullback = False
    if trend == "UP":
        dipped = any(c["low"] <= ema_slow + 2 * pip or c["low"] <= ema_fast for c in last4)
        pullback = dipped and last["close"] > ema_fast and last["close"] > last["open"]
    elif trend == "DOWN":
        spiked = any(c["high"] >= ema_slow - 2 * pip or c["high"] >= ema_fast for c in last4)
        pullback = spiked and last["close"] < ema_fast and last["close"] < last["open"]

    consecutive_bull = 0
    consecutive_bear = 0
    for i in range(n - 1, max(-1, n - 10) - 1, -1):
        if candles[i]["close"] > candles[i]["open"]:
            consecutive_bull += 1
        else:
            break
    for i in range(n - 1, max(-1, n - 10) - 1, -1):
        if candles[i]["close"] < candles[i]["open"]:
            consecutive_bear += 1
        else:
            break

    return {
        "trend": trend,
        "emaFast": ema_fast,
        "emaSlow": ema_slow,
        "swingTrend": swing_trend,
        "strength": min(100, strength),
        "atrPips": 0.0,  # filled by caller
        "consecutiveBull": consecutive_bull,
        "consecutiveBear": consecutive_bear,
        "pullback": pullback,
    }
